packetbeat/_meta/config/beat.reference.yml.tmpl

```diff
@@ -44,6 +44,20 @@ packetbeat.interfaces.internal_networks:
 # The default is 30 MB.
 #packetbeat.interfaces.buffer_size_mb: 30
 
+# To scale processing across multiple Packetbeat processes, a fanout group
+# identifier can be specified. When `fanout_group` is used the Linux kernel splits
+# packets across Packetbeat instances in the same group by using a flow hash. It
+# computes the flow hash modulo with the number of Packetbeat processes in order
+# to consistently route flows to the same Packetbeat instance.
+#
+# The value must be between 0 and 65535. By default, no value is set.
+#
+# This is only available on Linux and requires using `type: af_packet`. Each process
+# must be running in same network namespace. All processes must use the same
+# interface settings. You must take responsibility for running multiple instances
+# of Packetbeat.
```

Comment:
I'm wondering about the relative merits of running this as n×processes v n×sniffers in one process.

If we give the config an n rather than an ID, in `sniffer.New` the sniffer construction loop can do n sniffer constructions per device (fnv(devname+pid) for fanout id?) and then the user does not need to coordinate the process starts, nor make sure there is no id collision — though there should be a warning for this). The question is whether the OS scheduler will do a better job at distributing work than the runtime scheduler. Also, " The same should work for multiple TPacket objects within the same process." [in the docs](https://pkg.go.dev/github.com/google/gopacket/afpacket#TPacket.SetFanout) is probably less affirmative than I really like, though the mechanism is via fd, so there is no reason why it is not true.

Reply:
I'm torn on this as well. From a customer-deployment perspective, keeping everything contained in one process would be much simpler. The points that Dan brings up are valid, though.

Reply:
To start I totally think we should give consideration to allowing users to run concurrent sniffers on the same interface. My thinking with this change is that it does not prevent us from allowing this in the future. I'm glad you raised this because it will make sure we are thinking about how the config would work so that we don't do something that causes grief when we want to support n×sniffers.

> relative merits of running this as n×processes v n×sniffers in one process

When I was testing this I tried the approach of running multiple sniffers via config, but Packetbeat has a [restriction](https://github.com/elastic/beats/blob/1496ba72cc1d6f5ebfdb2b183132b8f55c4f66b7/packetbeat/config/config.go#L84) that prevents this.

```
packetbeat.interfaces:
- device: any
  fanout_group: 1
  type: af_packet
- device: any
  fanout_group: 1
  type: af_packet
```

> The question is whether the OS scheduler will do a better job at distributing work than the runtime scheduler.

I don't know the answer and it would be nice to have a recommendation. By allowing n×processes we are giving some flexibility for controlling process level settings like CPU affinity. Probably the runtime would do just as good a job, but I don't know.

> fnv(devname+pid) for fanout id?

I think an explicit fanout ID is better. If the user is operating any other AF_PACKET processes (like Zeek) then by having a predicable fanout ID they can ensure there are no collisions.


------

In summary, I think we can do an independent enhancement to support the n×sniffers model. The config would offer a `concurrency` option.

```
packetbeat.interfaces:
- device: any
  concurrency: 4 # Number of concurrent tasks to create for reading packets from the interface.
  fanout_group: 20001
  type: af_packet
```

Reply:
> > relative merits of running this as n×processes v n×sniffers in one process
> 
> When I was testing this I tried the approach of running multiple sniffers via config, but Packetbeat has a [restriction](https://github.com/elastic/beats/blob/1496ba72cc1d6f5ebfdb2b183132b8f55c4f66b7/packetbeat/config/config.go#L84) that prevents this.
> 
> ```
> packetbeat.interfaces:
> - device: any
>   fanout_group: 1
>   type: af_packet
> - device: any
>   fanout_group: 1
>   type: af_packet
> ```

I was thinking rather of having a numeric config something like

```
packetbeat.interfaces:
- device: any
  type: af_packet
  fanout: 2
  fanout_group_id: 1
```

This then constructs the fanout group and applies the policy to it.

> I don't know the answer and it would be nice to have a recommendation.

I don't have an answer.